Formatted output for cluster options (part 2)

nrwahl2 opened the pull request:
This is intended as a follow-up to #3323. That one can stand on its own, or we can scrap it and focus on this one.

The only remaining step is to add an option to a CLI tool, possibly `crm_attribute`. I hacked `scheduler_metadata()` to exercise the new API in both text and XML formats, and it looks reasonable.

Ref T620

nrwahl2 commented on tools/crm_attribute.c:
```diff
@@ -98,14 +98,31 @@ struct {
+    if (pcmk__str_eq(optarg, PCMK__VALUE_CLUSTER, pcmk__str_none)) {
+        options.list_type = pcmk__opt_table_cluster;
+    } else {
+        crm_warn("Invalid option type %s, using cluster options instead",
```
TODO: This warning doesn't work because logging hasn't been initialized until after option processing.

nrwahl2 commented on lib/common/options.c:
```diff
@@ -954,14 +954,30 @@ pcmk__cluster_option(GHashTable *options, const char *name)
+pcmk__format_option_metadata(pcmk__output_t *out, const char *name,
+                             const char *desc_short, const char *desc_long,
+                             enum pcmk__opt_context filter,
+                             enum pcmk__opt_table table)
```
slightly annoyed that I named the enum `pcmk__opt_table` while I'm usually calling the variables `option_list` or similar. Might rename. Similar in an informal sense, but we're used to `GHashTable`s and `GList`s.

nrwahl2 commented on lib/pacemaker/pcmk_options.c:
```diff
@@ -0,0 +1,41 @@
+    const char *desc_short = "Pacemaker cluster options";
+    const char *desc_long = NULL;
+
+    desc_long = "Also known as properties, these are options that affect "
```
arguably too verbose, but erred on the side of being friendly to newbies running list-options commands. not tied to it

kgaillot replied:
On second thought, ignore that. It could be a significant project on its own (created T776).

nrwahl2 replied:
The proposal in T776 seems better and more general, but it seems like it would be pretty straightforward to have `--long-desc` and `--short-desc` options that go with `--list-options`

kgaillot replied:
I'd rather not be stuck with maintaining them. The command-line option approach won't scale as we add more, and it's harder for higher-level tools to discover. We can live with the status quo.

nrwahl2 replied:
Agreed.

nrwahl2 commented on cts/cli/regression.tools.exp:
```diff
@@ -23,6 +23,585 @@ A new shadow instance was created. To begin using it, enter the following into y
+
+Also known as properties, these are options that affect behavior across the entire cluster. They are configured within cluster_property_set elements inside the crm_config subsection of the CIB configuration section.
+
+dc-version
```
This format is very much open to feedback

nrwahl2 replied:
I like getting rid of the short/long description headers -- the user doesn't care, and no one needs to parse it.

Putting the short description on the same line as the option name is pretty reasonable. It might look "off" for options that don't have a short desc... maybe in those cases put the long desc after the colon and omit the long desc line.

This is closer to the format of a fancy list. The main problem was that the way fancy lists are implemented, formatting it correctly for fancy would format it incorrectly for non-fancy. However, now I'm thinking we could just enable fancy (via a new internal function) unconditionally. This is new, so no one's relying on old text output. If I can get it working, that's probably the way to go.

Allowed values all on one line is probably better than separate lines. Again, no one is parsing this output.

I'm hesitant to overload "Allowed values" to list either a type or values, but it's probably fine. Listing a value type is meaningless for enumerated values (it's trivially a string type). And we can be less strict with the semantics of each field in text output compared to XML output.

Overall, yeah I think we should aim for something like this. I appreciate the idea. And I don't mind if this PR sits around for a little bit while we tweak the output, or if we revisit it later.

nrwahl2 replied:
Hmm what about translations? We could put the translation of short desc in parentheses at the end of the header line, and the translation of long desc after the English long desc on the indented line... or skip English altogether if a translation exists, though I doubt that's a good idea.

nrwahl2 replied:
Similarly, for the top-level short and long desc, we could put the translation in parentheses in the same paragraph or below it in a new paragraph...

kgaillot replied:
The point of translation is to use it instead of English. The layout would remain the same, it would just use the translation.

The only exception is OCF metadata XML, since OCF itself supports translations -- in that case, we always show the English, and add the locale's translation if available. You could make an argument that the XML output here should do the same, but the text output should show only the translation.

nrwahl2 replied:
That makes sense. I was going by the fact that we have historically included both languages (when output was always XML). What you're saying is easier and cleaner.

nrwahl2 commented:
Rebased on main and fixed Coverity-found issue

kgaillot commented on cts/cts-cli.in:
```diff
@@ -3354,6 +3370,7 @@ for t in $tests; do
         -e 's/last_change time=\".*\"/last_change time=\"\"/' \
         -e 's/ api-version="[^"]*"/ api-version="X"/' \
         -e 's/ default="[^"]*"/ default=""/' \
+        -e 's/ Default value: .*/ Default value:/' \
```
hmm, no great way around it ... maybe `s/ Default value: (val1|val2)/ Default value: OPT1_DEFAULT/` for known ones ...

nrwahl2 replied:
That's going to grow the sed significantly, and I'm not sure how much of a disadvantage there is to leaving it this way... a bit aggressive, but it'll affect only text output for whatever tools it hits.

Either way, the sed lines will have to be tweaked again depending on the text output format we settle on.

nrwahl2 replied:
Enumerating known ones feels like asking for trouble if they change in the future -- for example, if we add a new allowed value and a configure option to choose it as the default. We won't catch that `cts-cli` fails with that setting unless one of our CI machines uses that configure option.

kgaillot replied:
What I meant was enumerating just the values that *can* be configured differently. The idea being

    concurrent-fencing ...
        ...
        Allowed values: boolean (default: false)

or

    concurrent-fencing ...
        ...
        Allowed values: boolean (default: true)

would be replaced with

    concurrent-fencing ...
        ...
        Allowed values: boolean (default: CONCURRENT_FENCING_DEFAULT)

where CONCURRENT_FENCING_DEFAULT is a literal string and not a placeholder. But, we won't have the option name easily available, so probably just stripping defaults is the way to go.

kgaillot commented on po/zh_CN.po:
```diff
@@ -11,7 +11,7 @@ msgid ""
 "Project-Id-Version: Pacemaker 2\n"
 "Report-Msgid-Bugs-To: developers@clusterlabs.org\n"
-"POT-Creation-Date: 2024-02-01 16:24-0800\n"
+"POT-Creation-Date: 2024-02-07 20:47-0800\n"
```
don't bother with this commit -- if no actual strings changed, the line numbers aren't important until release time

nrwahl2 replied:
done

kgaillot reviewed (commented):
nice

nrwahl2 commented on daemons/fenced/pacemaker-fenced.c:
```diff
@@ -775,6 +775,14 @@ static pcmk__cluster_option_t fencer_options[] = {
+        pcmk__opt_context_none,
+        NULL,
+        NULL,
+    },
```
Right, was on the fence between that and being explicit. I tentatively did it this way based on a lot of our other NULL-terminated lists of structs.

I don't care either way

kgaillot replied:
No preference

nrwahl2 commented on lib/common/options.c:
```diff
@@ -852,73 +852,69 @@ pcmk__valid_placement_strategy(const char *value)
     char *new_value = NULL;
 
-    CRM_ASSERT(name != NULL);
+    CRM_ASSERT(option != NULL);
```
can't hurt

nrwahl2 commented on tools/crm_attribute.c:
```diff
@@ -105,6 +106,14 @@ struct {
+static gboolean
+list_cb(const gchar *option_name, const gchar *optarg, gpointer data,
+        GError **error) {
+    options.command = 'L';
```
probably trivial to do in this PR if we want (assuming it's local to this file)

kgaillot replied:
why not -- enum rsc_command in crm_resource.c is existing example

nrwahl2 replied:
done

nrwahl2 commented on cts/cli/regression.tools.exp:
```diff
@@ -23,6 +23,599 @@ A new shadow instance was created. To begin using it, enter the following into y
+    panic
+
+election-timeout
+  Short description: *** Advanced Use Only ***
```
The effort is not completely trivial but I think this could be worthwhile for user experience.

nrwahl2 commented on daemons/schedulerd/pacemaker-schedulerd.c:
```diff
@@ -112,14 +130,20 @@ main(int argc, char **argv)
+                g_set_error(&error, PCMK__EXITC_ERROR, exit_code,
+                            "Error creating output format xml: %s",
+                            pcmk_rc_str(rc));
+            }
```
done

nrwahl2 commented on include/crm/crm.h:
```diff
@@ -69,7 +69,7 @@ extern "C" {
  * >=3.19.0: DC supports PCMK__CIB_REQUEST_COMMIT_TRANSACT
  */
-#  define CRM_FEATURE_SET		"3.19.0"
+#  define CRM_FEATURE_SET		"3.19.1"
```
done

nrwahl2 commented:
Rebased on current main

nrwahl2 commented:
Updated to address most of review:
* Dropped NLS commit
* Dropped mention of C API from feature set bump commit message
* Added command enum to `crm_attribute.c`
* Added assertions of required option members
* Added `pcmk__daemon_metadata()` and used it for CIB manager, controller, and scheduler metadata. Check return codes. The fencer doesn't use this yet because its metadata doesn't live in libcrmcommon yet, but it does follow the same return code scheme.
* Changed the text output format to align with Ken's suggestions.
  * Still fuzzy about where to put translations. For now, top-level translations go in their own parenthesized paragraph, while option-level translations go on the same line as their English version.
  * The output is arguably a bit cramped. We could optionally add blank lines between options, or we could keep it compact.
  * I added a new `pcmk__output_text_set_fancy()` to set the fancy boolean, hid and deprecated the `--text-fancy` CLI option, and dropped the `pcmk__force_args()` calls that use it.

TODO:
* Flags for advanced and pacemaker-generated
* Tweak the `cts-cli` sed so that tests should pass regardless of `configure` options

nrwahl2 commented:
Implemented new option flags: `pcmk__opt_advanced`, `pcmk__opt_generated`, `pcmk__opt_deprecated`

nrwahl2 commented:
Main change this time is adding a schema... phew that was painful to figure out. Thanks @clumens for finding the `include` element though.

nrwahl2 commented on lib/common/options_display.c:
```diff
@@ -13,19 +13,197 @@
+    out->info(out, "%s", desc_short);
+#ifdef ENABLE_NLS
+    if (strcmp(desc_short, _(desc_short)) != 0) {
+        out->info(out, "(%s)", desc_short, _(desc_short));
```
wrong, but changing this output anyway

kgaillot commented on lib/common/options.c:
```diff
@@ -448,8 +448,7 @@ static pcmk__cluster_option_t cluster_options[] = {
         pcmk__opt_context_schedulerd,
-        N_("*** Deprecated *** "
-            "Whether to remove stopped resources from the executor"),
+        N_("Whether to remove stopped resources from the executor"),
```
The deprecated and advanced flags should also cause the `*** Deprecated ***`/`*** Advanced use only ***` labels to be shown in text output. In XML output, we should use OCF's `<deprecated/>` capability, and I guess a custom `advanced="true"` XML attribute

nrwahl2 replied:
That's reasonable. I omitted them to avoid clutter since (1) as of the most recent push we had stopped displaying deprecated options and (2) the user indicated they were fine with seeing advanced options since they specified `--all`.

We're bringing deprecated back though.

kgaillot replied:
I think if someone runs --all they still benefit from knowing which options merit more caution

nrwahl2 replied:
You're right

nrwahl2 replied:
Do you think these labels should precede the short description, or be their own bullet points before or after the long description?

nrwahl2 replied:
I lean toward first bullet point, above long desc...

kgaillot replied:
Some possibilities:

    * remove-after-stop: Whether to remove stopped resources from the executor
      * DEPRECATED
      * Values other than default are poorly tested and potentially dangerous. This option will be removed in a future release.
      * Allowed values: true, false (default)
    
    * remove-after-stop (DEPRECATED): Whether to remove stopped resources from the executor
      * Values other than default are poorly tested and potentially dangerous. This option will be removed in a future release.
      * Allowed values: true, false (default)
    
    * remove-after-stop: Whether to remove stopped resources from the executor (DEPRECATED)
      * Values other than default are poorly tested and potentially dangerous. This option will be removed in a future release.
      * Allowed values: true, false (default)
    
    * remove-after-stop: Whether to remove stopped resources from the executor
      * Values other than default are poorly tested and potentially dangerous. This option will be removed in a future release. (DEPRECATED)
      * Allowed values: true, false (default)

Alternatively, we could show all non-advanced, non-default options first, then advanced options (with a heading), then deprecated options (with a heading). I kind of like that idea.

kgaillot commented on lib/common/options_display.c:
```diff
@@ -59,6 +59,12 @@ static void
+    const char *generated = "";
+
+    if (pcmk_is_set(option->flags, pcmk__opt_generated)) {
+        generated = "(generated by Pacemaker) ";
```
`_()`

kgaillot commented on cts/cli/regression.tools.exp:
```diff
@@ -30,10 +30,10 @@ Also known as properties, these are options that affect behavior across the enti
   * dc-version: Pacemaker version on cluster node elected Designated Controller (DC)
     * Includes a hash which identifies the exact revision the code was built from. Used for diagnostic purposes.
-    * Allowed values: string (default: )
+    * Allowed values: (generated by Pacemaker) string (default: )
```
Maybe "Possible values (generated by Pacemaker): string (default: none)"?

nrwahl2 replied:
That's probably better. I don't like having `(generated by Pacemaker)` right in front of the option list, because it makes it look like it's associated with the first option. Likewise with putting it after the option list.

kgaillot reviewed (commented):
looking good

nrwahl2 commented on lib/common/options_display.c:
```diff
@@ -13,19 +13,197 @@
+    CRM_ASSERT((out != NULL) && (desc_short != NULL) && (desc_long != NULL)
+               && (option_list != NULL));
+
+    out->info(out, "%s", desc_short);
```
Nowhere. This is the top level -- a short description and long description of the option list.

The XML output has a name (for example, `"cluster-options"`) because that's required by OCF. I don't think it sheds much light on the text output, however, so I dropped the name from default output.

kgaillot replied:
Ah, I missed that. It's fine as is

nrwahl2 commented on lib/common/options_display.c:
```diff
@@ -13,19 +13,197 @@
+    if (strcmp(desc_long, _(desc_long)) != 0) {
+        out->info(out, "(%s)", desc_long, _(desc_long));
+    }
+#endif  // ENABLE_NLS
```
yep, thanks

nrwahl2 commented on lib/common/options_display.c:
```diff
@@ -13,19 +13,197 @@
+
+    } else {
+        out->list_item(out, "Allowed values", "%s (default: none)",
+                       option->type);
```
That works for me. It's already unambiguous in the sense that `none` isn't in quotes like the true defaults are -- but I think `no default` is clearer.

nrwahl2 commented on lib/common/options_display.c:
```diff
@@ -13,19 +13,197 @@
+        }
+        free(str);
+
+        out->list_item(out, "Allowed values", "%s", buf->str);
```
Any thoughts about `"%s (default: \"%s\")"`? Best I can come up with is
```
"%s " _("(default: ") "\"%s\""
```

nrwahl2 replied:
Welp:
```
options_display.c: In function 'add_allowed_values_default':
options_display.c:57:24: error: called object is not a function or function pointer
   57 |                        "%s (" _("default") ": \"%s\")",
      |                        ^~~~~~
options_display.c:57:44: error: expected ')' before string constant
   57 |                        "%s (" _("default") ": \"%s\")",
      |                                            ^~~~~~~~~~~
```

I could build it as a GString...

nrwahl2 commented on include/crm/common/options_internal.h:
```diff
@@ -77,10 +77,10 @@ typedef struct pcmk__cluster_option_s {
-                                   const char *desc_short,
-                                   const char *desc_long,
-                                   enum pcmk__opt_context filter);
+int pcmk__format_option_metadata(pcmk__output_t *out, const char *name,
```
Oh, duh.

nrwahl2 commented on lib/common/output_text.c:
```diff
@@ -336,6 +347,25 @@ pcmk__mk_text_output(char **argv) {
+    private_data_t *priv = NULL;
+
+    CRM_ASSERT((out != NULL)
+               && pcmk__str_eq(out->fmt_name, "text", pcmk__str_none));
```
I could do that... Most/all of the existing format-specific functions like this use `CRM_CHECK()`. See `pcmk__output_xml_create_parent()`, `pcmk__output_xml_add_node_copy()`, `pcmk__output_create_xml_node()`, `pcmk__output_{get,set}_log_level()`, `pcmk__output_set_log_filter()`, etc. I thought I was following convention (I thought it was `CRM_ASSERT()`).

With that said, I also meant to create a text-specific output function for `option-list` that wraps the default function and sets fancy. I don't think there's ever a reason to use non-fancy output for `option-list`. It would be borderline-unreadable.

kgaillot replied:
I don't think check/assert is necessary to set a global behavior option, it's more like "if this is xml, do it this way". (The log set level/filter functions could fit that category too.) Functions that modify the XML in a specific way make sense to check/assert since they don't have a use case with any other type of output object.

nrwahl2 replied:
Makes sense

nrwahl2 commented on cts/cli/regression.tools.exp:
```diff
@@ -23,6 +23,433 @@ A new shadow instance was created. To begin using it, enter the following into y
+    * Allowed values: string (default: )
+  * cluster-name: An arbitrary name for the cluster
+    * This optional value is mostly for users' convenience as desired in administration, but may also be used in Pacemaker configuration rules via the #cluster-name node attribute, and by higher-level tools and resource agents.
+    * Allowed values: string (default: )
```
Either way would be fine. I lean slightly toward blank lines, mostly because I like the idea of "Allowed values" being separated to make it easier to see at a glance.

pcs already does terminal-based line wrapping FWIW, though it will be pulling from our XML output first. It wouldn't hurt for us to do line wrapping for text output as a wishlist item. Mostly noting that users of high-level tools may already have that.

kgaillot replied:
Blank line would be fine

nrwahl2 commented on cts/cli/regression.tools.exp:
```diff
@@ -23,6 +23,433 @@ A new shadow instance was created. To begin using it, enter the following into y
+  </status>
+</pacemaker-result>
+=#=#=#= End test: List all available cluster options (invalid value) (XML) - Incorrect usage (64) =#=#=#=
+* Passed: crm_attribute  - List all available cluster options (invalid value) (XML)
```
Thank you!

nrwahl2 commented on lib/common/options_display.c:
```diff
@@ -343,7 +346,10 @@ option_list_xml(pcmk__output_t *out, va_list args)
-        if ((filter == pcmk__opt_context_none) || (filter == option->context)) {
+        if (!pcmk_is_set(option->flags, pcmk__opt_deprecated)
+            && ((filter == pcmk__opt_context_none)
+                || (filter == option->context))) {
```
seems to defeat the purpose; easy enough though

kgaillot replied:
Basically --all is for humans, XML is for parsers that don't care about information overload. If we had something like --all=true/false it would be fine to default text to false and XML to true (we do something similar with crm_mon).

nrwahl2 replied:
That makes sense and wouldn't be too hard, but probably not worth the complication.

nrwahl2 replied:
I could do `--all=(true|false|default)` at the `crm_attribute` level and require a boolean at the C API level. `crm_attribute` chooses appropriately if `--all=default`.

Not sure if I want to do that or just let XML show everything unconditionally. Lowest priority.

kgaillot commented on include/crm/common/options_internal.h:
```diff
@@ -36,6 +36,18 @@ bool pcmk__env_option_enabled(const char *daemon, const char *option);
+    pcmk__opt_advanced   = (1 << 0),    //!< Advanced use only
+    pcmk__opt_generated  = (1 << 1),    //!< Generated by Pacemaker
+    pcmk__opt_deprecated = (1 << 2),    //!< Deprecated
+};
```
We do want a single context per option, but it could make sense to ask for multiple contexts when displaying them

kgaillot commented on tools/crm_attribute.c:
```diff
@@ -393,6 +409,22 @@ delete_attr_on_node(xmlNode *child, void *userdata)
+        g_set_error(&error, PCMK__EXITC_ERROR, CRM_EX_USAGE,
+                    "Invalid --list-option value '%s'. Allowed values: "
+                    PCMK__VALUE_CLUSTER,
+                    options.opt_list);
```
If glib guarantees it, it's fine as is

nrwahl2 replied:
It guarantees it unless **we** screw up in the future and land ourselves in `command_list()`. That's not the callback function; it's what we call *after* option parsing.

nrwahl2 commented on lib/pacemaker/pcmk_options.c:
```diff
@@ -39,3 +41,23 @@ pcmk__list_cluster_options(pcmk__output_t *out)
+
+    pcmk__xml_output_finish(out, pcmk_rc2exitc(rc), xml);
+    return rc;
+}
```
Yeah probably... maybe I was thinking about reducing the amount of doxygen lol, I dunno. Or maybe `pcmk__output_option_metadata()` started out more complex. Or -- and I'm thinking this is the one -- maybe I didn't want to have a `pcmk__output_cluster_options()` in libcrmcommon and `pcmk__list_cluster_options()` in libpacemaker because they're so similar. (The output one has to stay in libcrmcommon for the sake of the options table.)

kgaillot replied:
It would be fine to have a high-level public API in libpacemaker and rely on the libcrmcommon function internally

nrwahl2 replied:
That's a given at this point. I'm trying to remember why I did the enum thing in the first place, and thinking it might have been to avoid having such similar names

nrwahl2 commented:
Have not done translation testing yet. About to have to go to class, so here's what I've got. I think this addresses all current review.

kgaillot commented:
needs a rebase on main

nrwahl2 commented:
Not sure why that was required; it didn't require any conflict resolution and the resulting number of commits is the same. Ah well, rebase done.

kgaillot commented on cts/cli/regression.tools.exp:
```diff
@@ -23,6 +23,481 @@ A new shadow instance was created. To begin using it, enter the following into y
+
+  * cluster-infrastructure: The messaging layer on which Pacemaker is currently running
+    * Used for informational and diagnostic purposes.
+    * Possible values (generated by Pacemaker): string (default: )
```
@tomjelinek @gao-yan , here's what the text output of the new crm_attribute option would look like. If you're not concerned with structured data, you could just call that and be done with it. (The regression tests strip the default values because they may depend on configure options. The actual output would show the defaults.)

kgaillot commented on cts/cli/regression.tools.exp:
```diff
@@ -23,6 +23,481 @@ A new shadow instance was created. To begin using it, enter the following into y
+        <longdesc lang="en">Includes a hash which identifies the exact revision the code was built from. Used for diagnostic purposes.</longdesc>
+        <shortdesc lang="en">Pacemaker version on cluster node elected Designated Controller (DC)</shortdesc>
+        <content type="string" default=""/>
+      </parameter>
```
@tomjelinek @gao-yan , and here's what the `--output=as=xml` output would look like. It's the same as the subdaemon metadata commands except it's wrapped in `<pacemaker-result>` (and it shows all the options).

kgaillot commented on lib/common/options_display.c:
```diff
@@ -130,8 +133,12 @@ option_list_default(pcmk__output_t *out, va_list args)
+    const bool all = (bool) va_arg(args, int);
 
-    bool old_fancy = false;
+    const bool old_fancy = false;
```
const snuck in this commit and got removed in the next

nrwahl2 replied:
Whoops, thanks

kgaillot commented on lib/pacemaker/pcmk_options.c:
```diff
@@ -38,3 +40,23 @@ pcmk__list_cluster_options(pcmk__output_t *out)
+
+// Documented in header
+int
+pcmk_list_cluster_options(xmlNode **xml)
```
This could take the place of the tmp_out usage

nrwahl2 replied:
You'd think so :/ But this doesn't take a filter argument (needed by `pcmk__daemon_metadata()`) or a set of name/desc_long/desc_short arguments (needed by `fencer_metadata()`).

nrwahl2 commented on cts/cli/regression.daemons.exp:
```diff
@@ -101,52 +101,52 @@
       </longdesc>
       <shortdesc lang="en">
-        *** Advanced Use Only ***
+        Declare an election failed if it is not decided within this much time. If you need to adjust this value, it probably indicates the presence of a bug.
```
> check for

All `pcs` or `crmsh` would have to do is look for `--list-options` in the output of `crm_attribute --help-all`. If and only if it's present, you can run `crm_attribute --list-options=cluster --output-as=xml` to get all the options in the new format instead of running the daemon metadata commands.

Whether you all have time to add parsing for the new format, is another matter.

> I'm guessing with current pcs code, we would just lose the indicators

It actually breaks entirely.
```
$ sudo pcs property describe
Error: Agent 'pacemaker-based' is not installed or does not provide valid metadata: Invalid attribute advanced for element parameter, line 12
Error: Agent 'pacemaker-controld' is not installed or does not provide valid metadata: Invalid attribute advanced for element parameter, line 12
Error: Agent 'pacemaker-schedulerd' is not installed or does not provide valid metadata: Invalid attribute advanced for element parameter, line 12
Error: Errors have occurred, therefore pcs is unable to continue
```

pcs is trying to validate the output against the OCF RA 1.1 schema. The OCF schema doesn't allow these custom attributes, which is why I've overridden a few pieces of it in the new `crm_attribute-2.34.rng` schema.

---

Incidentally, pcs's parsing and of allowed values is broken since 2.1.7. It very specifically depends on the presence of TWO spaces before "Allowed values" in the longdesc. But in 2.1.7, I removed one of the spaces as gratuitous, assuming it didn't matter.
```
lib/resource_agent/pcs_transform.py:        parts = parameter.longdesc.split("  Allowed values: ")
```

tomjelinek replied:
@kgaillot @nrwahl2:
Validating against modified OCF schema (`/usr/lib/pcs/data/ocf-1.1.rng`) is hard-coded and it cannot go away. The reason is that the real OCF schema has never been actually adhered to. To be able to at least somehow validate the xml structure, we had to come up with a relaxed version of the schema and bundle it with pcs.

Moving from plaintext to structured data for marking advanced parameters, allowed parameter values and deprecations is welcome. However, pcs needs to be compatible with both the new and the old way at the same time. We cannot be sure that upstream users match the new pcs with the new pacemaker. It would be easier to have OCF 1.2 standard to differentiate between the two approaches (plaintext vs. structure). (And to prevent further deviation from the OCF 1.1.) But I think we can manage it in pcs with OCF 1.1 as well, as long as there is either of plaintext or attribute specification and not both for the given metadata. If you want this to be done in 9.5, it is manageable. It would certainly help to finish the pacemaker side sooner than later, so we have enough time to get it done.

> Incidentally, pcs's parsing of allowed values is broken since 2.1.7. It very specifically depends on the presence of TWO spaces before "Allowed values" in the longdesc.

That's easy to fix. Just switch such parameters to `content type="select"` and put the allowed values to `option` elements, as specified in OCF 1.1 :wink: 

Moving to `crm_attribute --list-options=cluster` is another change on top of that, though. If the old way of running daemon metadata commands is preserved, no matter if it produces metadata in the current format or the one discussed here, I'm fine with that. It would give us time to switch to `crm_attribute`.

nrwahl2 replied:
@tomjelinek
> Validating against modified OCF schema (`/usr/lib/pcs/data/ocf-1.1.rng`) is hard-coded and it cannot go away.

All we'd have to do is update that schema to allow two new optional attributes in the `<parameter>` element: `advanced` and `generated`. Since they're optional, it wouldn't break on older Pacemakers.

> OCF 1.2 standard to differentiate between the two approaches

I have to defer to Ken on whether it's reasonable to have a new official OCF schema accept this output. The three differences between this cluster option output and the OCF 1.1 schema are:
* No `<actions>` element (required by OCF)
* Two option types (`percentage` and `time`) that are not allowed by OCF
* Two new `<parameter>` attributes (`advanced` and `generated`)

Depends on whether we want to relax the official schema in those ways. It *would* be backward compatible, since nothing that was previously accepted would become rejected.

> as long as there is either of plaintext or attribute specification and not both for the given metadata

The new XML output has *only* the structured (attribute) method. The plaintext strings are no longer in the descriptions.

The new text output still has plaintext indicators, but they're in separate bullet points instead of part of the description. So they're even less useful to parse than before :)

> finish the pacemaker side sooner than later

I'm hoping we can have it merged next week. The draft was done over a week ago; we've just been reviewing and updating since then. It seems like we're approaching a good place since you and gao were pinged.

> That's easy to fix. Just switch such parameters to `content type="select"` and put the allowed values to option elements, as specified in OCF 1.1 😉

For the most part, they already are and have been for some time now, but `pcs` grabs the "Allowed values: ..." from the `longdesc` instead. For example, in Pacemaker 2.1.7 (latest release):
```
    <parameter name="no-quorum-policy">
      <longdesc lang="en">What to do when the cluster does not have quorum  Allowed values: stop, freeze, ignore, demote, suicide</longdesc>
      <shortdesc lang="en">What to do when the cluster does not have quorum</shortdesc>
      <content type="select" default="">
        <option value="stop" />
        <option value="freeze" />
        <option value="ignore" />
        <option value="demote" />
        <option value="suicide" />
      </content>
    </parameter>
```

I recall there were a couple of options that *should* have been `select`-type with `<content>` and were not until recent commits, but most were.

And yes, it's easy ;) I was just pointing out that I introduced a regression in `pcs` by changing from two spaces to one.

---

> If the old way of running daemon metadata commands is preserved, no matter if it produces metadata in the current format or the one discussed here

In the current proposal, the old way of running daemon metadata commands is preserved, and it produces metadata in the format discussed here.

tomjelinek replied:
> > Validating against modified OCF schema (`/usr/lib/pcs/data/ocf-1.1.rng`) is hard-coded and it cannot go away.
> 
> All we'd have to do is update that schema to allow two new optional attributes in the `<parameter>` element: `advanced` and `generated`. Since they're optional, it wouldn't break on older Pacemakers.

Agreed, that's the plan. I probably should have stated it clearly: the modified OCF schema isn't going away, but it can be modified as needed.

> > OCF 1.2 standard to differentiate between the two approaches
> 
> I have to defer to Ken on whether it's reasonable to have a new official OCF schema accept this output.

Well, there are more points that would need to be addressed to finally make the OCF schema compatible with agents and pacemaker metadata. Like attributes used by fence agents, custom order of xml elements, percentage and time data types... I don't even remember all the bits which don't comply with the standard and we had to cover them in our custom schema. (It's easy to find them out, though - just take metadata of all agents and try to validate them.) It probably wouldn't pay off to make the new version just for this pacemaker metadata. But having a new version covering _all_ the things actually being used in agents and pacemaker metadata could be worth it.

> > as long as there is either of plaintext or attribute specification and not both for the given metadata
> 
> The new XML output has _only_ the structured (attribute) method. The plaintext strings are no longer in the descriptions.

:+1:

> > finish the pacemaker side sooner than later
> 
> I'm hoping we can have it merged next week. The draft was done over a week ago; we've just been reviewing and updating since then. It seems like we're approaching a good place since you and gao were pinged.

:+1:

> > That's easy to fix. Just switch such parameters to `content type="select"` and put the allowed values to option elements, as specified in OCF 1.1 😉
> 
> They already are. For the most part, they have been for a long time now, but `pcs` still grabs the "Allowed values: ..." from the `longdesc` instead.

Oh, ok. We'll take a look at that. Most probably we want to skip this plaintext parsing in case structured data are provided. We'll do the same for `advanced` and `generated`. That way, pcs would be able to process both the new and the old format.

> > If the old way of running daemon metadata commands is preserved, no matter if it produces metadata in the current format or the one discussed here
> 
> In the current proposal, the old way of running daemon metadata commands is preserved, and it produces metadata in the format discussed here.

:+1:

kgaillot replied:
> All `pcs` or `crmsh` would have to do is look for `--list-options` in the output of `crm_attribute --help-all`.

That's what the feature set is for. We don't want external code parsing help

> Well, there are more points that would need to be addressed to finally make the OCF schema compatible with agents and pacemaker metadata. Like attributes used by fence agents, custom order of xml elements, percentage and time data types... I don't even remember all the bits which don't comply with the standard and we had to cover them in our custom schema. (It's easy to find them out, though - just take metadata of all agents and try to validate them.) It probably wouldn't pay off to make the new version just for this pacemaker metadata. But having a new version covering _all_ the things actually being used in agents and pacemaker metadata could be worth it.

Yeah, the "Resource Agent API" column in https://projects.clusterlabs.org/project/view/37/ is collecting possible changes. However resource-agents hasn't even been updated for 1.1 yet, so it would be better to push that forward first.

kgaillot replied:
Actually this went into main for 2.1.8. The crm_attribute option is new, so it doesn't matter. The daemons will keep their metadata options for the 2.1 series (maybe even 3.0), and their output will remain the same except it will have the new advanced/generated/deprecated syntax.

kgaillot replied:
We could hack it so it keeps the output identical when called via the daemons, but it would be ugly

kgaillot replied:
@nrwahl2 , it sounds like it would be better to retrofit it to keep the daemon output identical. We could potentially overload legacy_xml for that purpose.

nrwahl2 replied:
> @nrwahl2 , it sounds like it would be better to retrofit it to keep the daemon output identical.

My thought was that if `advanced="1"`, the high-level tools could add "*** Advanced use only ***" or whatever marker they prefer...

> We could potentially overload legacy_xml for that purpose.

I don't think we should use that any more than we have to. We could add a `bool legacy` option to the new `option-list` message functions, setting it to `false` everywhere except `pcmk__daemon_metadata()`. Keep the legacy stuff clearly separated from the usual format so that it's easy to drop at 3.0.0 or whenever it's allowed.

kgaillot replied:
> > @nrwahl2 , it sounds like it would be better to retrofit it to keep the daemon output identical.
> 
> My thought was that if `advanced="1"`, the high-level tools could add "*** Advanced use only ***" or whatever marker they prefer...

Right, but we don't want to require higher-level tools (or user scripts etc.) to change for a minor-minor release. I was hoping they would ignore the new attributes and we'd just lose the markers, but they're actually looking for the markers and/or validating the schema, which means anyone building the newest Pacemaker and running with an existing pcs/crm would have problems.

> > We could potentially overload legacy_xml for that purpose.
> 
> I don't think we should use that any more than we have to. We could add a `bool legacy` option to the new `option-list` message functions, setting it to `false` everywhere except `pcmk__daemon_metadata()` and the fencer metadata. Keep the legacy stuff clearly separated from the usual format so that it's easy to drop at 3.0.0 or whenever it's allowed.

Actually that's why I was thinking of using legacy_xml -- we're already planning to drop that at 3.0, so this would require no extra work then. With legacy_xml we just drop the code that uses it; with a new argument, we'd ideally drop the argument meaning every caller has to be changed.

nrwahl2 replied:
> anyone building the newest Pacemaker and running with an existing pcs/crm would have problems.

Fair enough. I also discovered this morning that `pcs property set` fails now, even with `--force`.

> With legacy_xml we just drop the code that uses it; with a new argument, we'd ideally drop the argument meaning every caller has to be changed.

Makes sense. Honestly the amount of difference in work is negligible. Most of the work will probably be re-implementing the old format alongside the new one, which has to be done for either approach.
* To use `legacy_xml()`, we have to add a getter function so that the message functions can know whether `legacy_xml` is set. The message functions are the ones that would have to actually *use* it. The functions in `output_xml.c` themselves wouldn't do anything special for `option-list` specifically. And the daemon callers would have to set `legacy_xml`. Not a big deal but it's not nothing.
* Adding a new argument involves trivial changes to two or three message functions and only a small handful of callers.

Both of these are pretty easy to add or to drop.

kgaillot replied:
Either approach would be fine

nrwahl2 replied:
You win lol. The existence of a nested call chain (`pcmk__list_cluster_options()` and `pcmk__output_cluster_options()`), and the fact that we're going to be adding several more message function callers, makes me think overloading `legacy_xml` is the cleaner approach. It's hackier but we know that it's temporary. I don't foresee there being any weird interference between the two uses, since option listing and `crm_mon` should never overlap.

nrwahl2 commented on cts/cli/regression.tools.exp:
```diff
@@ -23,6 +23,481 @@ A new shadow instance was created. To begin using it, enter the following into y
+  * placement-strategy: How the cluster should allocate resources to nodes
+    * Possible values: "default" (default), "utilization", "minimal", "balanced"
+
+  * Advanced options:
```
Good point. The top one is prettier but as you mentioned, the bottom one is the most convenient by far.

nrwahl2 commented on cts/cli/regression.tools.exp:
```diff
@@ -23,6 +23,463 @@ A new shadow instance was created. To begin using it, enter the following into y
+    * A cluster node may receive notification of a "succeeded" fencing that targeted it if fencing is misconfigured, or if fabric fencing is in use that doesn't cut cluster communication. Use "stop" to attempt to immediately stop Pacemaker and stay stopped, or "panic" to attempt to immediately reboot the local node, falling back to stop on failure.
+    * Possible values: "stop" (default), "panic"
+
+  * election-timeout: Declare an election failed if it is not decided within this much time. If you need to adjust this value, it probably indicates the presence of a bug.
```
Well, these aren't being filtered out like they should be. Neither is the deprecated option (`remove-after-stop`). They're not even separated unless `--all` is given.

clumens commented on daemons/fenced/pacemaker-fenced.c:
```diff
@@ -788,12 +776,29 @@ fencer_metadata(void)
+    }
+
+    out->message(tmp_out, "option-list", name, desc_short, desc_long,
+                 pcmk__opt_none, fencer_options, true);
```
I ran this PR through my formatted output arg checking, and aside from me having to change something, this was the one line that causes problems:
```pacemaker-fenced.c:789:5: error: Expected 'uint32_t', but got 'int' in argument 6
  789 |     out->message(tmp_out, "option-list", name, desc_short, desc_long,
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  790 |                  pcmk__opt_none, fencer_options, true);
      |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Casting `pcmk__opt_none` to `uint32_t` fixes this.  I'm not sure I want to do anything about this in my tool since gcc expects an unsigned type but thinks it's getting a signed one.

nrwahl2 replied:
That would be reasonable to fix on the caller side. In `pcmk__daemon_metadata()`, I already cast an `enum pcmk__opt_flags` to `uint32_t`.

nrwahl2 replied:
I did change this `out->message` call, but in hindsight it's pretty nonstandard to do so. We usually **don't** cast our enum values to `uint32_t` before passing them as flag group arguments.

kgaillot replied:
Defining `pcmk__opt_none` to `0U` should take care of it (or `UINT32_C(0)`, but I don't think that's necessary in this case)

nrwahl2 replied:
Good idea, I'll do that as well as `(1U << x)` for the other flags

nrwahl2 commented:
Just realized "true" and "false" aren't allowed in the OCF schema, because that would make way too much sense.

Updating to change to "1" and "0".

fabbione commented:
retest this please

nrwahl2 commented on lib/pacemaker/pcmk_options.c:
```diff
@@ -0,0 +1,40 @@
+
+    desc_long = "Also known as properties, these are options that affect "
+                "behavior across the entire cluster. They are configured "
+                "within " PCMK_XE_CLUSTER_PROPERTY_SET " elements inside the "
```
Are the string constants going to be a problem for translations? I just remembered that when this eventually gets passed to the `"option-list"` message function, it will be used as `_(desc_long)`.

kgaillot replied:
Yes, that is an annoying limitation

fabbione commented:
retest this please

nrwahl2 commented:
Refactored the new schema a bit. Instead of having `crm_attribute-2.34.rng` override `ocf-ra-1.1.rng`, I added an `options-2.34.rng` API schema to do the overrides. The `crm_attribute` schema then pulls in `options-2.34.rng` as an `externalRef` in the top-level definition.

This will avoid duplication in the next PR when we start adding options to `crm_resource`.

fabbione commented:
retest this please

fabbione commented:
retest this please

fabbione commented:
retest this please

fabbione commented:
retest this please

fabbione commented:
retest this please

fabbione commented:
retest this please

nrwahl2 commented:
Just addressed the two remaining review points. About to rebase on main.

nrwahl2 commented:
yet another rebase on main

fabbione commented:
retest this please

kgaillot commented on include/crm/common/options_internal.h:
```diff
@@ -46,6 +46,9 @@ enum pcmk__opt_flags {
     pcmk__opt_schedulerd = (1U << 2),   //!< In scheduler metadata
+    pcmk__opt_advanced   = (1U << 3),   //!< Advanced use only
+    pcmk__opt_generated  = (1U << 4),   //!< Generated by Pacemaker
+    pcmk__opt_deprecated = (1U << 5),   //!< Deprecated
```
Just for clarity :) maybe "Option is deprecated"

kgaillot commented on lib/common/options_display.c:
```diff
@@ -141,12 +151,50 @@ option_list_default(pcmk__output_t *out, va_list args)
+        advanced = g_slist_reverse(advanced);
+
+        out->spacer(out);
+        out->begin_list(out, NULL, NULL, "Advanced options");
```
`_()`

kgaillot commented on lib/common/options_display.c:
```diff
@@ -141,12 +151,50 @@ option_list_default(pcmk__output_t *out, va_list args)
+        deprecated = g_slist_reverse(deprecated);
+
+        out->spacer(out);
+        out->begin_list(out, NULL, NULL, "Deprecated options");
```
`_()`

kgaillot commented on tools/crm_attribute.c:
```diff
@@ -393,6 +409,22 @@ delete_attr_on_node(xmlNode *child, void *userdata)
+        // @TODO Improve usage messages to reduce duplication
+        exit_code = CRM_EX_USAGE;
+        g_set_error(&error, PCMK__EXITC_ERROR, CRM_EX_USAGE,
+                    "Invalid --list-option value '%s'. Allowed values: "
```
options

kgaillot commented on lib/common/output_log.c:
```diff
@@ -349,34 +349,56 @@ pcmk__mk_log_output(char **argv) {
 
-    priv = out->priv;
-    return priv->log_level;
+    if (pcmk__str_eq(out->fmt_name, "log", pcmk__str_none)) {
```
also move priv assert inside here

kgaillot commented on lib/common/output_log.c:
```diff
@@ -349,34 +349,56 @@ pcmk__mk_log_output(char **argv) {
-    CRM_ASSERT(out != NULL && out->priv != NULL);
-    CRM_CHECK(pcmk__str_eq(out->fmt_name, "log", pcmk__str_none), return);
+    if (pcmk__str_eq(out->fmt_name, "log", pcmk__str_none)) {
+        private_data_t *priv = out->priv;
```
ditto

kgaillot commented on lib/common/output_log.c:
```diff
@@ -390,14 +412,14 @@ void
-    priv->function = function;
-    priv->line = line;
-    priv->tags = tags;
+    if (pcmk__str_eq(out->fmt_name, "log", pcmk__str_none)) {
```
ditto

kgaillot reviewed (commented):
last touches

nrwahl2 commented on lib/common/output_text.c:
```diff
@@ -336,6 +350,50 @@ pcmk__mk_text_output(char **argv) {
+        private_data_t *priv = out->priv;
+
+        return priv->fancy;
+    }
```
Then this needs to be done in the log output helpers too

kgaillot reviewed (approved):
Woohoo!
